### ISSUE TITLE
C++ GC + polymorphism

### ISSUE THREAD

**Issue by soywiz** [jtransc/jtransc#124]
I have typed C++ code, but now it crashes. I think that it is due to polymorphism.
I'm going to try to use `p_java_lang_Object` in arguments and return types and proper types inside functions for now to see if that way it doesn't crash.

**Comment by soywiz** [comment]
@intrigus I don't know how to debug this, so I have added a boolean `CppGenerator.ENABLE_TYPING` to toggle it so you can check the generated code. Maybe you can figure out why is it crashing so we can fix it.
Ideally it should be typed to get the best performance possible.

**Comment by soywiz** [comment]
Random thought:

Instead of

```
struct java_lang_Object : public gc {}
struct java_lang_String : public java_lang_Object {}
```

-->

Could this work? And do `new final_java_lang_String` but casting to `base_`. Maybe that way the GC knows how to handle each object?
```
struct base_java_lang_Object {}
struct base_java_lang_String : public base_java_lang_Object {}

struct final_java_lang_Object : base_java_lang_Object , public gc {}
struct final_java_lang_String : base_java_lang_String , public gc {}
```

Do you know how robovm does this?

**Comment by intrigus** [comment]
Do you have a failing test case for me? Or does any test crash?

**Comment by soywiz** [comment]
HelloWorld test crashes in ArrayList.addAll if you toggle ENABLE_TYPING

**Comment by intrigus** [comment]
Will look into it, when I'm back home.
But currently this p_java_lang_Object approach looks horrible and it don't understand the purpose of it.
Will hopefully understand it when I'm on my pc :)

**Comment by soywiz** [comment]
I had problems using `java_lang_Object*` (and others `java_...*`) when tried yesterday to type it.  Not sure why, but creating a typedef per type, allowed me to compile it at least.

I think in C, this is different:

```
java_lang_Object* a, b, c; // a is a pointer, while b and c not.
```

```
typedef java_lang_Object* p_java_lang_Object;
p_java_lang_Object a, b, c; // a, b, c are all pointers
```

I think I had problems related to that but not exclusively.

Also we should probably use this so the GC is precise instead of conservative which will prevent some leaks with random data:

http://www.hpl.hp.com/hosted/linux/mail-archives/gc/2004-January/000281.html
```
GC_gcj_malloc was really designed for the GNU static Java compiler.
It assumes that the first word of every object contains a pointer
to a method table (or other type descriptor) and the second word
of that is a "GC descriptor" which identifies the possible pointer
locations in the object.  (The details are somewhat adjustable, but only with
extreme caution.  The thread interactions are tricky.)
```

**Comment by intrigus** [comment]
Yeah, we should really use GC_gcj_malloc but not now. I'd rather have good support for threads, atomics and jni. And then we could "fix" the GC.

**Comment by soywiz** [comment]
Sure. I agree with you that there are other priorities first that should be addressed. So no rush here, just wanted to put here as a remember for the future :)

**Comment by soywiz** [comment]
I think I'm approaching to the issue. I think that it is related that interfaces are not related to `java.lang.Object`, just like this:

```c++
struct O {
};

struct I1 {
};

struct B : O, I1 {
};

struct C : O, I1 {
};

int main() {
        B* b = new B();
        I1* i1 = dynamic_cast<I1*>(b);
        O* o = dynamic_cast<O*>(i1); // Problem here

        return 0;
}
```

So maybe we should add to interfaces a virtual method to get the java_lang_Object pointer of the object.

**Comment by soywiz** [comment]
This will be required (at least partially) for #127 . This works. And fails without the virtual operator. This way we can cast from interfaces to java_lang_Object

```
#include <stdio.h>

struct A { // java_lang_Object
};

struct B { // interface
        virtual operator A*() = 0;
};

struct C : public A, public B { // another class extending object and implementing interface
        virtual operator A*() { return dynamic_cast<A*>(this); }
};

int main() {
        C* c = new C();
        B* b = dynamic_cast<B*>(c);
        A* a = dynamic_cast<A*>(b);
        printf("HELLO\n");
        printf("%p\n", a);
        printf("%p\n", b);
        printf("%p\n", c);
        return 0;
}
```
